File: api/device_list.py
#!/usr/bin/env python3

"""
Copyright (c) 2021  Christoph Willing, Brisbane Australia
SPDX-License-Identifier: MIT
"""

import sys, os
import paho.mqtt.client as mqtt
import json


def main():

	def on_connect(client, userdata, flags, rc):
		print("Connected with result code "+str(rc))
		client.subscribe("/xkeys/server/#")

	def on_subscribe(client, userdata, mid, granted_qos):
		client.publish('/xkeys/node', '{"request":"deviceList"}')

	def on_message(client, userdata, msg):
		print(json.dumps(json.loads(msg.payload), indent=2))
		sys.exit(0)

	client = mqtt.Client()
	client.on_connect = on_connect
	client.on_message = on_message
	client.on_subscribe = on_subscribe

	client.connect("localhost", 1883, 60)

	client.loop_forever()


if __name__ == '__main__':
    main()

File: api/device_list.js
#!/usr/bin/env node

/*
Copyright (c) 2021  Christoph Willing, Brisbane Australia
SPDX-License-Identifier: MIT
*/


const mqtt = require('../node_modules/mqtt')


const connectUrl = 'mqtt://localhost';
const qos = 0;

var client = mqtt.connect(connectUrl);
client.on('error', (error) => {
	console.log('Connection failed:', error)
})
client.on('connect', () => {
	console.log('connected')
	//client.subscribe({'/xkeys/server/xkeys_button/#':{qos:qos}}, function (err, granted) {
	client.subscribe({'/xkeys/server/#':{qos:qos}}, function (err, granted) {
		if (!err) {
			console.log("Subscribed OK, granted: " + JSON.stringify(granted));
			client.publish('/xkeys/node', '{"request":"deviceList"}')
		} else {
			console.log('Subscription failed: ' + err)
		}
	})
})
client.on('close', () => {
	console.log("connection closed");
})

client.on('message', (topic, message) => {
	var message_obj = "";
	try {
		message_obj = JSON.parse(message);
		if (message_obj.request == "result_deviceList") {
			console.log("Device List:");
			console.log(message_obj.data);
			process.exit(0);
		} else {
			console.log('Received unhandled msg: ' + message_obj.request);
		}
	}
	catch (e) {
		console.log('ERROR parsing message: ' + e);
	}
})

